api/v1beta1/nutanixmachine_types.go

```diff
@@ -109,6 +141,84 @@ type NutanixMachineSpec struct {
 	GPUs []NutanixGPU `json:"gpus,omitempty"`
 }
 
+// NutanixMachineVMDisk defines the disk configuration for a NutanixMachine
+type NutanixMachineVMDisk struct {
+	// diskSize is the size (in Quantity format) of the disk attached to the VM.
+	// See https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#Format for the Quantity format and example documentation.
+	// The minimum diskSize is 1GB.
+	// +kubebuilder:validation:Required
+	DiskSize resource.Quantity `json:"diskSize"`
+
+	// deviceProperties are the properties of the disk device.
+	// +optional
+	DeviceProperties *NutanixMachineVMDiskDeviceProperties `json:"deviceProperties,omitempty"`
+
+	// storageConfig are the storage configuration parameters of the VM disks.
+	// +optional
+	StorageConfig *NutanixMachineVMStorageConfig `json:"storageConfig,omitempty"`
+
+	// dataSource refers to a data source image for the VM disk.
+	// +optional
+	DataSource *NutanixResourceIdentifier `json:"dataSource,omitempty"`
+}
+
+// NutanixMachineVMDiskDeviceProperties defines the device properties for a NutanixMachineVMDisk
+type NutanixMachineVMDiskDeviceProperties struct {
+	// deviceType specifies the disk device type.
+	// The valid values are "Disk" and "CDRom", and the default is "Disk".
+	// +kubebuilder:default=Disk
+	// +kubebuilder:validation:Required
+	DeviceType NutanixMachineDiskDeviceType `json:"deviceType"`
+
+	// adapterType is the adapter type of the disk address.
+	// If the deviceType is "Disk", the valid adapterType can be "SCSI", "IDE", "PCI", "SATA" or "SPAPR".
+	// If the deviceType is "CDRom", the valid adapterType can be "IDE" or "SATA".
+	// +kubebuilder:validation:Required
+	AdapterType NutanixMachineDiskAdapterType `json:"adapterType,omitempty"`
+
+	// deviceIndex is the index of the disk address. The valid values are non-negative integers, with the default value 0.
+	// For a Machine VM, the deviceIndex for the disks with the same deviceType.adapterType combination should
+	// start from 0 and increase consecutively afterwards. Note that for each Machine VM, the Disk.SCSI.0
+	// and CDRom.IDE.0 are reserved to be used by the VM's system. So for dataDisks of Disk.SCSI and CDRom.IDE,
+	// the deviceIndex should start from 1.
+	// +kubebuilder:default=0
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	DeviceIndex int32 `json:"deviceIndex,omitempty"`
+}
+
+// NutanixMachineVMStorageConfig defines the storage configuration for a NutanixMachineVMDisk
+type NutanixMachineVMStorageConfig struct {
+	// diskMode specifies the disk mode.
+	// The valid values are Standard and Flash, and the default is Standard.
+	// +kubebuilder:default=Standard
+	DiskMode NutanixMachineDiskMode `json:"diskMode"`
+
+	// storageContainer refers to the storage_container used by the VM disk.
+	// +optional
+	StorageContainer *NutanixStorageContainerResourceIdentifier `json:"storageContainer"`
```

Comment:
Can we just use standard resource identifier type here?

Reply:
I really prefer to avoid usage of NutanixResourceIdentifier here, because in v3 it still difficult to get storage container by name and required PE creds. On other side if we'll make it with NutanixResourceIdentifier and disable to provide name type, so I think we'll confuse our users

Reply:
See this https://github.com/mesosphere/konvoy2/blob/main/pkg/nutanix/client/storage.go#L14-L87

Reply:
@thunderboltsid Is this nutanix/client repo code for prism-v3 or v4?

Reply:
v3 at the moment because v4 APIs weren't GA when it was written.

Reply:
Ok, I've changed storage container to use NutanixResourceIdentifier.